packages/flutter/lib/src/material/colors.dart

```diff
@@ -88,6 +90,133 @@ class MaterialAccentColor extends ColorSwatch<int> {
+///   2. Use [MaterialStateColor.resolveWith] and pass in a callback that
+///      will be used to resolve the color in the given states.
+///
+/// This should only be used as parameters when they are documented to take
+/// [MaterialStatefulColor], otherwise only the default state will be used.
+///
+/// Example:
+/// ```dart
+/// Color getTextColor(Set<MaterialSet> states) {
+///   final interactiveStates = <MaterialState>{
+///     MaterialState.pressed,
+///     MaterialState.hovered,
+///     MaterialState.focused,
+///   };
+///   if(states.any(interactiveStates.contains) {
+///     return Colors.blue[900];
+///   }
+///   return Colors.blue[600];
+/// }
+///
+/// FlatButton(
+///   ...
+///   textColor: MaterialStateColor.resolveWith(getTextColor),
+/// ),
+/// ```
+abstract class MaterialStateColor extends Color {
+  /// Creates a [MaterialStateColor].
+  ///
+  /// If you want a `const` [MaterialStateColor], you'll need to extend
+  /// [MaterialStateColor] and override the [resolve] method. You'll also need
+  /// to provide a `defaultValue` to the super constructor, so that we can know
+  /// at compile-time what the value of the default [Color] is.
+  ///
+  /// Example (`const`):
+  ///
+  /// ```dart
+  /// class TextColor extends MaterialStateColor {
+  ///   static const int _defaultColor = 0xcafefeed;
+  ///   static const int _pressedColor = 0xdeadbeef;
+  ///
+  ///   const TextColor() : super(_defaultColor);
+  ///
+  ///   @override
+  ///   Color resolve(Set<MaterialState> states) {
+  ///     if (states.contains(MaterialState.pressed)) {
+  ///       return const Color(_pressedColor);
+  ///     }
+  ///     return const Color(_defaultColor);
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// If you don't need your [MaterialStateColor] to be `const`, consider using
+  /// [MaterialStateColor.resolveWith].
+  const MaterialStateColor(int defaultValue) : super(defaultValue);
+
+  /// Creates a [MaterialStateColor] from a [ColorResolver] callback function.
+  ///
+  /// If used as a regular color, the color resolved in the default state will
+  /// be used.
+  ///
+  /// The given callback parameter must return a non-null color in the default
+  /// state (empty set).
+  factory MaterialStateColor.resolveWith(ColorResolver callback) => _MaterialStateColor(callback);
+
+  /// This method resolves to a [Color] based on a given set of states.
+  ///
+  /// Example:
+  ///
+  /// ```dart
+  /// @override
+  /// Color resolve(Set<MaterialState> states) {
+  ///   if (states.contains(MaterialState.pressed)) {
+  ///     return Colors.grey[900];
+  ///   }
+  ///   return Colors.grey[600];
+  /// }
+  /// ```
+  Color resolve(Set<MaterialState> states);
+
+  /// Returns the color for the given set of states if `color` is a
+  /// [MaterialStateColor], otherwise returns the color itself.
+  ///
+  /// This is useful for widgets that accept [Color] params, this will work for
```

Comment:
... for widgets that have parameters which can be [Color] or [MaterialStateColor] values.

Reply:
Done.